packages/aws_signature_v4/lib/src/credentials/credentials_provider.dart

```diff
@@ -0,0 +1,118 @@
+/// Exception thrown when AWS credentials are either unavailable or invalid.
+/// {@endtemplate}
+class InvalidCredentialsException implements Exception {
+  /// {@macro aws_signature_v4.invalid_credentials_exception}
+  const InvalidCredentialsException([this.message]);
+
+  /// Exception thrown when AWS credentials could not be loaded by an
+  /// [AWSCredentialsProvider].
+  const InvalidCredentialsException.couldNotLoad()
+      : this('Could not load credentials from environment');
+
+  /// Further information about the exception, if any.
+  final String? message;
+
+  @override
+  String toString() =>
+      'InvalidCredentialsException${message == null ? '' : ': $message'}';
+}
+
+/// A utility for retrieving AWS credentials at runtime.
+abstract class AWSCredentialsProvider {
+  /// Creates a [StaticCredentialsProvider] with the given [credentials].
+  const factory AWSCredentialsProvider(AWSCredentials credentials) =
+      StaticCredentialsProvider;
+
+  /// {@template aws_signature_v4.environment_credentials_provider}
+  /// Creates an [AWSCredentialsProvider] for credentials injected
+  /// via the platform-specific environment.
+  ///
+  /// On the VM, this uses `Platform.environment`. On the Web, this uses the
+  /// Dart environment, e.g. `--dart-define` for Flutter.
+  ///
+  /// To always use the Dart environment, use [AWSCredentialsProvider.dartEnvironment]
+  /// instead.
+  /// {@endtemplate}
+  const factory AWSCredentialsProvider.environment() =
+      EnvironmentCredentialsProvider;
+
+  /// {@template aws_signature_v4.dart_environment_credentials_provider}
+  /// Creates an [AWSCredentialsProvider] for credentials injected via the Dart
+  /// environment.
+  /// {@endtemplate}
+  const factory AWSCredentialsProvider.dartEnvironment() =
+      DartEnvironmentCredentialsProvider;
+
+  const AWSCredentialsProvider._();
+
+  /// Retrieves AWS credentials.
+  FutureOr<AWSCredentials> retrieve();
+}
+
+/// {@template aws_signature_v4.static_credentials_provider}
+/// Creates an [AWSCredentialsProvider] for a set of static, compile-time AWS
+/// credentials.
+/// {@endtemplate}
+class StaticCredentialsProvider extends AWSCredentialsProvider {
+  /// {@macro aws_signature_v4.static_credentials_provider}
+  const StaticCredentialsProvider(this._credentials) : super._();
+
+  final AWSCredentials _credentials;
+
+  @override
+  AWSCredentials retrieve() => _credentials;
+}
+
+/// {@macro aws_signature_v4.dart_environment_credentials_provider}
+class DartEnvironmentCredentialsProvider extends AWSCredentialsProvider {
+  /// {@macro aws_signature_v4.dart_environment_credentials_provider}
+  const DartEnvironmentCredentialsProvider() : super._();
+
+  @override
+  AWSCredentials retrieve() {
+    const accessKeyId = String.fromEnvironment($awsAccessKeyId);
+    const secretAccessKey = String.fromEnvironment($awsSecretAccessKey);
+    const sessionToken = String.fromEnvironment($awsSessionToken);
```

Comment:
Can these be `null`?

Reply:
No `String.fromEnvironment` has a default value of empty string